runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java

```diff
@@ -65,11 +77,23 @@ public static TestSparkRunner fromOptions(PipelineOptions options) {
     return new TestSparkRunner(sparkOptions);
   }
 
+  /**
+   * Overrides for the test runner.
+   */
+  @SuppressWarnings("unchecked")
   @Override
-  public <OutputT extends POutput, InputT extends PInput>
-      OutputT apply(PTransform<InputT, OutputT> transform, InputT input) {
-    return delegate.apply(transform, input);
-  };
+  public <OutputT extends POutput, InputT extends PInput> OutputT apply(
+          PTransform<InputT, OutputT> transform, InputT input) {
+    // if the pipeline is required to execute as a streaming pipeline,
+    // and the source is an adapted unbounded source (as bounded),
+    // read it as unbounded source via UnboundedReadFromBoundedSource.
+    if (isStreaming && transform instanceof BoundedReadFromUnboundedSource) {
```

Comment:
I generally agree with with going even more towards portability, but doesn't losing `PipelineRunner.apply` mean losing the runner's ability to manipulate the `Pipeline` ? Of course the user's `Pipeline` shouldn't change logically, but runners should have the freedom to replace transforms to achieve the execution of the "same" logic, no ?